### PR TITLE
feat(rnd,blocks): Add D-ID Block

### PR DESCRIPTION
**Pull request by aarushik93**
### Background

Adding D-ID block

### Changes 🏗️

Added docs to assist finding available voices
Added md description linking to voice docs
Added target="_blank" to force opening in new tab when clicking link in descriptions

### PR Quality Scorecard ✨

- [x] Have you used the PR description template? &ensp; `+2 pts`
- [ ] Is your pull request atomic, focusing on a single change? &ensp; `+5 pts`
- [ ] Have you linked the GitHub issue(s) that this PR addresses? &ensp; `+5 pts`
- [ ] Have you documented your changes clearly and comprehensively? &ensp; `+5 pts`
- [ ] Have you changed or added a feature? &ensp; `-4 pts`
  - [ ] Have you added/updated corresponding documentation? &ensp; `+4 pts`
  - [ ] Have you added/updated corresponding integration tests? &ensp; `+5 pts`
- [ ] Have you changed the behavior of AutoGPT? &ensp; `-5 pts`
  - [ ] Have you also run `agbenchmark` to verify that these changes do not regress performance? &ensp; `+10 pts`

___

### **PR Type**
enhancement, dependencies

___

### **Description**
- Added a new `ClipsBlock` class to integrate with D-ID for creating video clips, including input and output schemas, and methods for clip creation and status polling.
- Enhanced error handling for the clip creation process to manage failures and timeouts.
- Updated the `Secrets` class to include a new field for storing the D-ID API key.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>talking_head.py</strong><dd><code>Add D-ID video clip creation block</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

rnd/autogpt_server/autogpt_server/blocks/talking_head.py

<li>Introduced <code>ClipsBlock</code> class for D-ID video clip integration.<br> <li> Defined input and output schemas for video clip creation.<br> <li> Implemented methods to create clips and poll for their status.<br> <li> Added error handling for clip creation process.<br>

</details>

  </td>
  <td><a href="https://github.com/Significant-Gravitas/AutoGPT/pull/7798/files#diff-9531dfc1d5aac327e1eccf45ca5629b5147d295e045b0b590f7c25c89f5c606a">+147/-0</a>&nbsp; </td>

</tr>                    
</table></td></tr><tr><td><strong>Dependencies</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>settings.py</strong><dd><code>Add D-ID API key to secrets configuration</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

rnd/autogpt_server/autogpt_server/util/settings.py

- Added `did_api_key` field to `Secrets` class.

</details>

  </td>
  <td><a href="https://github.com/Significant-Gravitas/AutoGPT/pull/7798/files#diff-281666fe9783cb96474f35739db160eceb136aca260e20e73990edc91f19610a">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

File: rnd/autogpt_server/autogpt_server/blocks/talking_head.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
+import time
+from typing import Literal
+
+import requests
+
+from autogpt_server.data.block import Block, BlockCategory, BlockOutput, BlockSchema
+from autogpt_server.data.model import BlockSecret, SchemaField, SecretField
+
+
+class ClipsBlock(Block):
```
Comment (Pwuts) [inline rnd/autogpt_server/autogpt_server/blocks/talking_head.py]: pls more descriptive name :)

File: rnd/autogpt_server/autogpt_server/blocks/talking_head.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+        subtitles: bool = SchemaField(
+            description="Whether to include subtitles", default=False
+        )
+        ssml: bool = SchemaField(description="Whether the input is SSML", default=False)
+        max_polling_attempts: int = SchemaField(
+            description="Maximum number of polling attempts", default=30
+        )
+        polling_interval: int = SchemaField(
+            description="Interval between polling attempts in seconds", default=10
+        )
```
Comment (Pwuts) [inline rnd/autogpt_server/autogpt_server/blocks/talking_head.py]: I'd advise adding minimum values to these two fields

File: rnd/autogpt_server/autogpt_server/blocks/talking_head.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+        script_input: str = SchemaField(
+            description="The text input for the script", default="Welcome to AutoGPT"
+        )
+        provider: Literal["microsoft", "elevenlabs", "amazon"] = SchemaField(
+            description="The voice provider to use", default="microsoft"
+        )
+        voice_id: str = SchemaField(
+            description="The voice ID to use", default="en-US-JennyNeural"
+        )
```
Comment (Pwuts) [inline rnd/autogpt_server/autogpt_server/blocks/talking_head.py]: maybe add a (markdown) link to the description pointing to the docs w/ a list of available voices or how to find a voice ID?

File: rnd/autogpt_server/autogpt_server/blocks/talking_head.py (hunk 1 above)
Comment (Swiftyos) [reply rnd/autogpt_server/autogpt_server/blocks/talking_head.py]: I'd second this, the builder is currently struggling with unclear names. It is currently the only thing the are searched on so needs to be super clear.

File: rnd/autogpt_server/autogpt_server/util/settings.py [hunk 4]
```diff
@@ -108,6 +108,7 @@ class Secrets(UpdateTrackingModel["Secrets"], BaseSettings):
 
     medium_api_key: str = Field(default="", description="Medium API key")
     medium_author_id: str = Field(default="", description="Medium author ID")
+    did_api_key: str = Field(default="", description="D-ID API Key")
```
Comment (majdyz) [inline rnd/autogpt_server/autogpt_server/util/settings.py:111]: Can we add this into .env.example ?

**Review by majdyz — DISMISSED** [review]

File: rnd/autogpt_server/autogpt_server/util/settings.py (hunk 4 above)
Comment (aarushik93) [reply rnd/autogpt_server/autogpt_server/util/settings.py:111]: updated!

**Comment by aarushik93** [comment]
Gonna merge (as we discussed, once getting a review and only changing small things, we can just force merge for speed's sake!)

**Pull request merged**